flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResources.java

```diff
@@ -0,0 +1,162 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.declarative;
+
+import org.apache.flink.api.common.JobStatus;
+import org.apache.flink.runtime.executiongraph.ArchivedExecutionGraph;
+import org.apache.flink.runtime.executiongraph.ExecutionGraph;
+import org.apache.flink.runtime.jobmaster.slotpool.ResourceCounter;
+
+import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
+
+import java.time.Duration;
+
+/**
+ * State which describes that the scheduler is waiting for resources in order to execute the job.
+ */
+class WaitingForResources implements State, ResourceConsumer {
+
+    private final Context context;
+
+    private final Logger logger;
+
+    private final ResourceCounter desiredResources;
+
+    WaitingForResources(Context context, Logger logger, ResourceCounter desiredResources) {
+        this.context = context;
+        this.logger = logger;
+        this.desiredResources = desiredResources;
+    }
+
+    @Override
+    public void onEnter() {
+        context.runIfState(this, this::resourceTimeout, Duration.ofSeconds(10L));
+        notifyNewResourcesAvailable();
+    }
+
+    @Override
+    public void cancel() {
+        context.goToFinished(context.getArchivedExecutionGraph(JobStatus.CANCELED, null));
+    }
+
+    @Override
+    public void suspend(Throwable cause) {
+        context.goToFinished(context.getArchivedExecutionGraph(JobStatus.SUSPENDED, cause));
+    }
+
+    @Override
+    public JobStatus getJobStatus() {
+        return JobStatus.INITIALIZING;
+    }
+
+    @Override
+    public ArchivedExecutionGraph getJob() {
+        return context.getArchivedExecutionGraph(getJobStatus(), null);
+    }
+
+    @Override
+    public void handleGlobalFailure(Throwable cause) {
+        context.goToFinished(context.getArchivedExecutionGraph(JobStatus.FAILED, cause));
+    }
+
+    @Override
+    public Logger getLogger() {
+        return logger;
+    }
+
+    @Override
+    public void notifyNewResourcesAvailable() {
+        if (context.hasEnoughResources(desiredResources)) {
+            createExecutionGraphWithAvailableResources();
+        }
+    }
+
+    private void resourceTimeout() {
+        createExecutionGraphWithAvailableResources();
+    }
+
+    private void createExecutionGraphWithAvailableResources() {
+        try {
+            final ExecutionGraph executionGraph =
+                    context.createExecutionGraphWithAvailableResources();
+
+            context.goToExecuting(executionGraph);
+        } catch (Exception exception) {
+            logger.error("handling initialization failure", exception);
```

Comment:
We should clean this up a bit. I only added it back then because it seems like this transition was never logged, but ideally the context logs all state transitions with appropriate the exception.